File: app/src/main/java/com/nimtego/plectrum/presentation/di/modules/data/RepositoryModule.java
package com.nimtego.plectrum.presentation.di.modules.data;

import android.content.Context;

import com.nimtego.plectrum.App;
import com.nimtego.plectrum.data.cache.Cache;
import com.nimtego.plectrum.data.cache.CacheK;
import com.nimtego.plectrum.data.cache.DashBoardEntityCache;
import com.nimtego.plectrum.data.cache.FileManager;
import com.nimtego.plectrum.data.cache.Serializer;
import com.nimtego.plectrum.data.entity.mapper.EntityDataMapper;
import com.nimtego.plectrum.data.executor.BaseExecutor;
import com.nimtego.plectrum.data.model.rss_itunes.Feed;
import com.nimtego.plectrum.data.model.rss_itunes.PopularResponse;
import com.nimtego.plectrum.data.repository.repository.AppRepository;
import com.nimtego.plectrum.data.repository.datasource.DataStoreFactory;
import com.nimtego.plectrum.data.repository.repository.DashBoardRepository;
import com.nimtego.plectrum.domain.executor.ThreadExecutor;
import com.nimtego.plectrum.domain.repository.Repository;
import com.nimtego.plectrum.domain.repository.RepositoryK;
import com.nimtego.plectrum.presentation.di.modules.ActivityModule;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module(includes = ActivityModule.class)
public class RepositoryModule {

    @Provides
    @Singleton
    RepositoryK provideDashBoardRepository(EntityDataMapper mapper,
                                                   DataStoreFactory<PopularResponse> dataStoreFactory) {
        return new DashBoardRepository(dataStoreFactory, mapper);
    }

    @Provides
    EntityDataMapper entityDataMapper() {
        return new EntityDataMapper();
    }

    @Provides
    Serializer serializer() {
        return new Serializer();
    }

    @Provides
    FileManager fileManager() {
        return new FileManager();
    }

    @Provides
    CacheK<PopularResponse> cache(Context appContext,
                                        Serializer serializer,
                                        FileManager fileManager,
                                        ThreadExecutor threadExecutor) {
        return new DashBoardEntityCache<>(appContext, serializer, fileManager, threadExecutor);
    }

    @Provides
    DataStoreFactory<PopularResponse> storeFactory(Context appContext, CacheK<PopularResponse> cache) {
        return new DataStoreFactory<>(appContext, cache);
    }

    @Provides
    ThreadExecutor executor () {
        return new BaseExecutor();
    }

}

File: app/src/main/java/com/nimtego/plectrum/presentation/di/modules/data/RepositoryModule.kt
package com.nimtego.plectrum.presentation.di.modules.data

import com.nimtego.plectrum.data.model.mappers.*
import com.nimtego.plectrum.data.repository.datasource.detail.DetailStoreFactory
import com.nimtego.plectrum.data.repository.datasource.popular.book.PopularBookFactory
import com.nimtego.plectrum.data.repository.datasource.popular.movie.PopularMovieFactory
import com.nimtego.plectrum.data.repository.datasource.popular.music.PopularMusicFactory
import com.nimtego.plectrum.data.repository.datasource.search.SongDataStoreFactory
import com.nimtego.plectrum.data.repository.repository.*
import com.nimtego.plectrum.data.repository.repository.detail.MusicalDetail
import com.nimtego.plectrum.domain.repository.RepositoryPopular
import com.nimtego.plectrum.domain.repository.detail.MusicalDetailRepository
import com.nimtego.plectrum.presentation.di.modules.ContextModule
import com.nimtego.plectrum.presentation.di.modules.domain.RepositoryQualifiers
import com.nimtego.plectrum.presentation.mvp.model.main_tab_model.BaseParentViewModel
import com.nimtego.plectrum.presentation.mvp.model.main_tab_model.ChildViewModel
import dagger.Module
import dagger.Provides
import javax.inject.Named
import javax.inject.Singleton

@Module(includes = [ContextModule::class, DataStoreModule::class])
class RepositoryModule {


    @Provides
    @Singleton
    @Named(RepositoryQualifiers.MUSIC_REPOSITORY)
    internal fun providePopularMusicRepository(
            mapper: PopularMusicMapper,
            @Named(RepositoryQualifiers.MUSIC_REPOSITORY)
            dataStoreFactory: PopularMusicFactory
    ) : RepositoryPopular<BaseParentViewModel<ChildViewModel>> {
        return PopularMusicRepository(dataStoreFactory, mapper)
    }

    @Provides
    @Singleton
    @Named(RepositoryQualifiers.MUSIC_REPOSITORY)
    internal fun providePopularMovieRepository(
            mapper: PopularMovieMapper,
            @Named(RepositoryQualifiers.MOVIE_REPOSITORY)
            dataStoreFactory: PopularMovieFactory
    ) : RepositoryPopular<BaseParentViewModel<ChildViewModel>> {
        return PopularMovieRepository(dataStoreFactory, mapper)
    }

    @Provides
    @Singleton
    @Named(RepositoryQualifiers.MUSIC_REPOSITORY)
    internal fun providePopularBookRepository(
            mapper: PopularBookMapper,
            @Named(RepositoryQualifiers.MOVIE_REPOSITORY)
            dataStoreFactory: PopularBookFactory
    ) : RepositoryPopular<BaseParentViewModel<ChildViewModel>> {
        return PopularBookRepository(dataStoreFactory, mapper)
    }

    @Provides
    @Singleton
    @Named(RepositoryQualifiers.MUSIC_REPOSITORY)
    internal fun provideSongRepository(
            mapper: MusicalContentMapper,
            @Named(RepositoryQualifiers.MOVIE_REPOSITORY)
            dataStoreFactory: SongDataStoreFactory
    ) : SongRepository {
        return SongRepository(dataStoreFactory, mapper)
    }


    @Provides
    @Singleton
    internal fun provideMoreSectionRepository(mapperMusic: PopularMusicMapper,
                                              mapperMovie: PopularMovieMapper,
                                              mapperBook: PopularBookMapper,
                                              musicDataStoreFactory: PopularMusicFactory,
                                              movieDataStoreFactory: PopularMovieFactory,
                                              bookDataStoreFactory: PopularBookFactory) =
            MoreSectionRepository(musicDataStoreFactory = musicDataStoreFactory,
                                  movieDataStoreFactory = movieDataStoreFactory,
                                  bookDataStoreFactory = bookDataStoreFactory,
                                  mapperMusic = mapperMusic,
                                  mapperMovie = mapperMovie,
                                  mapperBook = mapperBook)

    @Provides
    @Singleton
    internal fun provideSearchRepository(mapper: MusicalContentMapper,
                                              dataStoreFactory: SongDataStoreFactory
    ): MusicalRepository {
        return MusicalRepository(dataStoreFactory, mapper)
    }

    @Provides
    @Singleton
    internal fun provideMusicalDetailRepository(
            mapper: MusicalDetailMapper,
            dataStoreFactory: DetailStoreFactory
    ): MusicalDetailRepository {
        return MusicalDetail(dataStoreFactory, mapper)
    }

    @Provides
    internal fun provideDetailMusicalMapper(): MusicalDetailMapper {
        return MusicalDetailMapper()
    }

    @Provides
    internal fun providePopularMusicMapper(): PopularMusicMapper {
        return PopularMusicMapper()
    }

    @Provides
    internal fun provideSearchMusicMapper(): MusicalContentMapper {
        return MusicalContentMapper()
    }

    @Provides
    internal fun providePopularMovieMapper(): PopularMovieMapper {
        return PopularMovieMapper()
    }

    @Provides
    internal fun providePopularBookMapper(): PopularBookMapper {
        return PopularBookMapper()
    }

}